scripts/mempool.js

```diff
@@ -0,0 +1,248 @@
+"use strict";
+
+/** An Unspent Transaction Output, used as Inputs of future transactions */
+class UTXO {
+    /**
+     * @param {Object} UTXO
+     * @param {String} UTXO.id - Transaction ID
+     * @param {String} UTXO.path - If applicable, the HD Path of the owning address
+     * @param {Number} UTXO.sats - Satoshi value in this UTXO
+     * @param {String} UTXO.script - HEX encoded spending script
+     * @param {Number} UTXO.vout - Output position of this transaction
+     * @param {Number} UTXO.height - Block height of the UTXO
+     * @param {Number} UTXO.status - UTXO status enum state
+     */
+    constructor({id, path, sats, script, vout, height, status} = {}) {
+        /** Transaction ID
+         * @type {String} */
+        this.id = id;
+
+        /** HD Path of the owning address
+         *  @type {String} */
+        this.path = path;
+
+        /** Satoshi value in this UTXO
+         *  @type {Number} */
+        this.sats = sats;
+
+        /** HEX encoded spending script
+         *  @type {String} */
+        this.script = script;
+
+        /** Output position of this transaction
+         *  @type {Number} */
+        this.vout = vout;
+
+        /** Block height of the UTXO
+         *  @type {Number} */
+        this.height = height;
+
+        /** UTXO status enum state
+         *  @type {Number} */
+        this.status = status;
+    }
+};
+
+/** A Mempool instance, stores and handles UTXO data for the wallet */
+class Mempool {
+    constructor() {
+        /** 
+         * An array of all known UTXOs
+         * @type {Array<UTXO>}
+         */
+        this.UTXOs = [];
```

Comment:
I think using Map would be more appropriate for the type of operations we're doing

Reply:
I was thinking mapping UTXO->status

Reply:
I don't see any real benefits of changing from array to map

Reply:
I'm feeling the same as @panleone, not sure if I'm missing anything though.

Reply:
Many functions are "Find the UTXO and change the status/remove it", in an array it's an O(n) operation, in a map it would be an average of O(1)

Reply:
it wouldn't be O(1) anyways because in many cases you want to get all the UTXOs with a fixed status

Reply:
Ok, with everything said I think changing to a map would outweigh the benefits, even if we had a lot of transactions it wouldn't be the bottleneck